fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java

```diff
@@ -346,14 +352,52 @@ protected Optional<ExpressionInfo> buildJoinExpr(GroupInfo leftGroup, GroupInfo
         pushRequiredColumns(leftExprInfo, leftExpression);
         pushRequiredColumns(rightExprInfo, rightExpression);
 
+        UKFKConstraints.JoinProperty joinProperty = null;
+        SessionVariable sessionVariable = ConnectContext.get().getSessionVariable();
+        if (sessionVariable.isEnableUKFKOpt()) {
+            UKFKConstraintsCollector.collectColumnConstraints(leftExprInfo.expr);
+            UKFKConstraintsCollector.collectColumnConstraints(rightExprInfo.expr);
+            UKFKConstraints constraint = UKFKConstraintsCollector.buildJoinColumnConstraint(newJoin,
+                    newJoin.getJoinType(), newJoin.getOnPredicate(), leftExprInfo.expr, rightExprInfo.expr);
+            joinProperty = constraint.getJoinProperty();
+        }
+
         // use small table as right child
         OptExpression joinExpr;
-        if (leftExprInfo.rowCount < rightExprInfo.rowCount) {
-            joinExpr = OptExpression.create(newJoin, rightExprInfo.expr,
-                    leftExprInfo.expr);
+        boolean reverse = false;
+        if (joinProperty != null && joinProperty.ukConstraint.isIntact) {
+            if (joinProperty.isLeftUK) {
+                if (allowFKAsRightTable(joinProperty, leftExprInfo, rightExprInfo)) {
+                    // Use the fk table as the right table
+                    joinExpr = OptExpression.create(newJoin, leftExprInfo.expr,
+                            rightExprInfo.expr);
+                } else {
+                    // If the uk table is too small or the fk table is too large, then use it as right table
+                    reverse = true;
```

Comment:
use uk table as right table?

Reply:
yes, put the fk table as the right table will increase the exchange cost. If the ratio of rk/uk is large enough, treat it as normal join.